__main__.py

```diff
@@ -0,0 +1,136 @@
+        }
+    
+    def episode_start(self, config: Dict[str, Any]):
+        """ Called at the start of each episode """
+        self.cartpole.reset(
+            config.get("initial_cart_position") or 0,
+            config.get("initial_pole_angle") or 0,
+            config.get("target_pole_position") or 0,
+        )
+
+    def episode_step(self, action: Dict[str, Any]):
+        """ Called for each step of the episode """
+        self.cartpole.step(action.get("command") or 0)
+
+    def halted(self) -> bool:
+        """ Should return True if the simulator cannot continue"""
+        # If the pole has fallen past 45 degrees, there's no use 
+        # in continuing.
+        return abs(self.cartpole._pole_angle) >= math.pi / 4
+
+if __name__ == "__main__":
+    # Grab standardized way to interact with sim API
+    sim = TemplateSimulatorSession()
+
+    # Configure client to interact with Bonsai service
+    config_client = BonsaiClientConfig()
+    client = BonsaiClient(config_client)
+
+    # Load json file as simulator integration config type file
+    with open('cartpole_interface.json') as file:
+        interface = json.load(file)
+
+    # Create simulator session and init sequence id
+    registration_info = SimulatorInterface(
+                            name=interface['name'], 
+                            timeout=interface['timeout'], 
+                            simulator_context=config_client.simulator_context, 
+    )
+    registered_session = client.session.create(
+                            workspace_name=config_client.workspace, 
+                            body=registration_info
+    )
+    print("Registered simulator.")
+    sequence_id = 1
+
+    try:
+        while True:
+            # Advance by the new state depending on the event type
+            sim_state = SimulatorState(
+                            session_id=registered_session.session_id,
+                            sequence_id=sequence_id, state=sim.get_state(), 
+                            halted=sim.halted()
+            )
+            event = client.session.advance(
+                        workspace_name=config_client.workspace, 
+                        session_id=registered_session.session_id, body=sim_state
+            )
+            sequence_id = event.sequence_id
+            print("[{}] Last Event: {}".format(time.strftime('%H:%M:%S'), 
+                                               event.type))
+
+            # Event loop
+            if event.type == 'Idle':
+                time.sleep(event.idle.callback_time)
+                print('Idling...')
+            elif event.type == 'EpisodeStart':
+                sim.episode_start(event.episode_start.config)
+            elif event.type == 'EpisodeStep':
+                sim.episode_step(event.episode_step.action)
+            elif event.type == 'EpisodeFinish':
+                print('Episode Finishing...')
+            elif event.type == 'unregister':
```

Comment:
`unregister` should be `Unregister`

Reply:
good catch. That's where the benefit of `EventTypesEventType` , but I currently find this much more readable for a first walkthrough. We could definitely benefit from an advanced walkthrough to show the rest of the API